pymc_experimental/utils/linear_cg.py

```diff
@@ -0,0 +1,279 @@
+            precond_residual = np.copy(residual)
+
+            linear_cg_retvals = linear_cg_updates(
+                result,
+                alpha,
+                residual_inner_prod,
+                eps,
+                beta,
+                residual,
+                precond_residual,
+                curr_conjugate_vec,
+            )
+            (
+                result,
+                alpha,
+                residual_inner_prod,
+                eps,
+                beta,
+                residual,
+                precond_residual,
+                curr_conjugate_vec,
+            ) = linear_cg_retvals
+
+        residual_norm = np.linalg.norm(residual, 2)
+        residual_norm = masked_fill(residual_norm, mask=rhs_is_zero, fill_value=0)
+        has_converged = residual_norm < stop_updating_after
+
+        if (
+            k >= min(10, max_iter - 1)
+            and bool(residual_norm.mean() < tolerance)
+            and not (n_tridiag and k < min(n_tridiag_iter, max_iter - 1))
+        ):
+            tolerance_reached = True
+            break
+
+        # Update tridiagonal matrices, if applicable
+        if n_tridiag and k < n_tridiag_iter and update_tridiag:
+            alpha_tridiag = np.copy(alpha)
+            beta_tridiag = np.copy(beta)
+
+            alpha_tridiag_is_zero = alpha_tridiag == 0
+            alpha_tridiag = masked_fill(alpha_tridiag, mask=alpha_tridiag_is_zero, fill_value=1)
+            alpha_reciprocal = 1 / alpha_tridiag
+            alpha_tridiag = masked_fill(alpha_tridiag, mask=alpha_tridiag_is_zero, fill_value=0)
+
+            if k == 0:
+                t_mat[k, k] = alpha_reciprocal
+            else:
+                t_mat[k, k] += np.squeeze(alpha_reciprocal + prev_beta * prev_alpha_reciprocal)
+                t_mat[k, k - 1] = np.sqrt(prev_beta) * prev_alpha_reciprocal
+                t_mat[k - 1, k] = np.copy(t_mat[k, k - 1])
+
+                if t_mat[k - 1, k].max() < 1e-6:
+                    update_tridiag = False
+
+            last_tridiag_iter = k
+
+            prev_alpha_reciprocal = np.copy(alpha_reciprocal)
+            prev_beta = np.copy(beta_tridiag)
+
+    # Un-normalize
+    result = result * rhs_norm
+    if not tolerance_reached and n_iter > 0:
+        raise RuntimeError(
+            "CG terminated in {} iterations with average residual norm {}"
+            " which is larger than the tolerance of {} specified by"
+            " gpytorch.settings.cg_tolerance."
+            " If performance is affected, consider raising the maximum number of CG iterations by running code in"
+            " a gpytorch.settings.max_cg_iterations(value) context.".format(
+                k + 1, residual_norm.mean(), tolerance
+            )
+        )
+
+    if n_tridiag:
+        t_mat = t_mat[: last_tridiag_iter + 1, : last_tridiag_iter + 1]
+        return result, t_mat.transpose(-1, *range(2, 2 + len(batch_shape)), 0, 1)
+    else:
+        return result
```

Comment:
This function returns two different types, depending on inputs.  This can make downstream code more complicated, since it will have to know how to handle both cases.  Is it possible to refactor this into two functions, one for the tridaigonal case and one not?

Reply:
Would it make sense to return `None` as the second argument when `n_tridiag` is False ?

Reply:
Checking the type vs checking the number of outputs isn't effectively very different.  If it doesn't make sense to split this, I think it makes sense to leave this as is and be mindful of this later, maybe return to it once we see how it fits in.

Reply:
I now return identity matrices if n_tridiag = 0. This ensures downstream computations based on tridiagonal matrices are not affected (Eigen values of identity matrices are all one and log of that would be zero, so the log determinant terms would be zero) see. equation S7 (https://arxiv.org/pdf/1809.11165.pdf)

The identity matrices assume n_tridiag = 1 to ensure dimensions are consistent for downstream computations.